### PR TITLE
Vendor btree

### PR DESCRIPTION
**Pull request by tgruben**
prevent allocs from occurring on PutConainerValues if container exists

### CODE REVIEW

File: roaring/assembly_test.go [hunk 1]
```diff
@@ -42,6 +42,7 @@ func TestBSFQ_CompareGo(t *testing.T) {
 		}
 	*/
 }
+var Result uint64
```
Comment (travisturner) [inline roaring/assembly_test.go]: does `Result` get returned somewhere?

File: roaring/containers_slice.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+func (sc *SliceContainers) Put(key uint64, c *container) {
+	i := search64(sc.keys, key)
+
+	// If index is negative then there's not an exact match
+	// and a container needs to be added.
+	if i < 0 {
+		sc.insertAt(key, c, -i-1)
+	} else {
+		//should this happen?
```
Comment (travisturner) [inline roaring/containers_slice.go]: it seems like it should. why the question?

File: roaring/containers_btree.go [hunk 3]
```diff
@@ -47,23 +47,31 @@ func (btc *BTreeContainers) Put(key uint64, c *container) {
...
-			oldV.n = n
-			oldV.mapped = mapped
-			return oldV, true
-		}
-		return &container{
-			containerType: containerType,
-			n:             n,
-			mapped:        mapped,
-		}, true
+type updater struct {
```
Comment (travisturner) [inline roaring/containers_btree.go]: just curious, does it help to have this `updater` struct? or is it just for code organizational purposes?
(i.e. is there anything functionally different besides returning `false` in the "exists" path?)

File: roaring/assembly_test.go (hunk 1 above)
Comment (tgruben) [reply roaring/assembly_test.go]: this prevents the optimizer from removing the code that the benchmark is supposed to test

File: roaring/containers_btree.go (hunk 3 above)
Comment (tgruben) [reply roaring/containers_btree.go]: yeah, this struct is used to hint to the compiler that the closure should just allocate on the stack,   without it escapes out to the heap

File: roaring/containers_slice.go (hunk 2 above)
Comment (tgruben) [reply roaring/containers_slice.go]: yeah, that should be removed.  It was just a note to myself as i was thinking

File: fragment.go [hunk 4]
```diff
@@ -189,7 +189,7 @@ func (f *Fragment) Open() error {
 func (f *Fragment) openStorage() error {
 	// Create a roaring bitmap to serve as storage for the slice.
 	if f.storage == nil {
-		f.storage = roaring.NewBitmap()
+		f.storage = roaring.NewBitmapBtree()
```
Comment (travisturner) [inline fragment.go]: For consistency with the other methods, can we rename this to `NewBTreeBitmap` (note also the capitalized `T` in tree)

File: roaring/assembly_test.go (hunk 1 above)
Comment (travisturner) [reply roaring/assembly_test.go]: I suggest adding a comment in the code with this information.

File: roaring/containers_btree.go (hunk 3 above)
Comment (travisturner) [reply roaring/containers_btree.go]: add a comment with this information

File: roaring/containers_test.go [hunk 5]
```diff
@@ -5,7 +5,8 @@ import (
 )
 
 func TestContainersIterator(t *testing.T) {
-	btc := NewBTreeContainers()
+	//btc := NewBTreeContainers()
```
Comment (travisturner) [inline roaring/containers_test.go]: can we modify the tests so that both container implementations get tested?

**Review by travisturner — CHANGES_REQUESTED** [review]

**Review by travisturner — APPROVED** [review]

**Pull request merged**